dace/frontend/vscode/vscode.py

```diff
@@ -0,0 +1,246 @@
+        s.connect((HOST, int(PORT)))
+        s.sendall(data_bytes)
+
+
+def recv(data: json):
+    """ Sends a json object to the port given as the env variable DACE_port
+        and returns the received json object
+        If the port isn't set we don't send anything.
+        :param data: json object with request data
+        :return: returns the object received
+    """
+
+    if not is_available():
+        return
+
+    HOST = socket.gethostname()
+    PORT = os.environ["DACE_port"]
+
+    data_bytes = bytes(json.dumps(data), "utf-8")
+    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
+        s.connect((HOST, int(PORT)))
+        s.sendall(data_bytes)
+        result = s.recv(1024)
+        return json.loads(result)
+    return None
+
+
+def send_bp_recv(data: json) -> json:
+    """ Sends a json object to the port given as the env variable DACE_port.
+        After the data is sent a breakpint is set. As soon as the debugger
+        continues, return the received data
+        :param data: json object with request data
+        :return: the received data as a json object
+    """
+
+    if not is_available():
+        return
+
+    HOST = socket.gethostname()
+    PORT = os.environ["DACE_port"]
+
+    data_bytes = bytes(json.dumps(data), "utf-8")
+    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
+        s.connect((HOST, int(PORT)))
+        s.sendall(data_bytes)
+        result = s.recv(1024)
+        return json.loads(result)
+    return None
+
+
+def sdfg_remove_instrumentations(sdfg: dace.sdfg.SDFG):
+    sdfg.instrument = dace.dtypes.InstrumentationType.No_Instrumentation
+    for state in sdfg.nodes():
+        state.instrument = dace.dtypes.InstrumentationType.No_Instrumentation
+        for node in state.nodes():
+            node.instrument = dace.dtypes.InstrumentationType.No_Instrumentation
+            if isinstance(node, dace.sdfg.nodes.NestedSDFG):
+                sdfg_remove_instrumentations(node.sdfg)
+
+
+def sdfg_has_instrumentation(sdfg: dace.sdfg.SDFG) -> bool:
+    if sdfg.instrument != dace.dtypes.InstrumentationType.No_Instrumentation:
+        return True
+    for state in sdfg.nodes():
+        if state.instrument != dace.dtypes.InstrumentationType.No_Instrumentation:
+            return True
+        for node in state.nodes():
+            if (hasattr(node, 'instrument') and node.instrument !=
+                    dace.dtypes.InstrumentationType.No_Instrumentation):
+                return True
+            if isinstance(node, dace.sdfg.nodes.NestedSDFG):
+                return sdfg_has_instrumentation(node.sdfg)
+    return False
+
+
+def AN_of_Array(graph, array_name: str,
+                state_id: int) -> [dace.sdfg.nodes.AccessNode, int]:
+    for i, node in enumerate(graph.nodes()):
+        if isinstance(node, dace.sdfg.nodes.AccessNode):
+            if node.label == array_name:
+                return (node, i)
```

Comment:
is it really the node we want? Are you looking for the topological-first or last node in the state instead?